### DISCUSSION TITLE
Add a custom model for forecasting time series

### DISCUSSION THREAD

**Discussion by Paandaman** [autogluon/autogluon#3091] [Q&A] [answered]
Autogluon tabular has a guide for how to add custom models but there is no such guide for time serie models. However, the time serie models are all based on AbstractTimeSeriesModel which itself inherits from AbstractModel, which is the same class that all tabular models are based on. 

So my question is, can I add a custom model in a similar way to how e.g. [the naive model](https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/local/naive.py) is implemented? If I follow the structure of AbstractTimeSeriesModel, will I be able to use that model in an ensemble together with the already existing models?

Many thanks in advance!

**Comment by shchur** [comment, marked as answer]
Hi, yes, this should be rather straightforward. Here is what you need to do:
1. Implement the model as a class inheriting from `AbstractTimeSeriesModel` and implement the `_fit` and `_predict` methods.
2. Modify`presets.py` :
    1. [Import the model class](https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/presets.py#L8)
    2. [Add model to the `MODEL_TYPES` dict](https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/presets.py#L27)
    3. (Optionally) Specify `DEFAULT_MODEL_PRIORITY` and add the model to the [presets dict](https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/presets.py#L105)

If you did 1, 2i, 2ii, you can train the model as
```python
predictor.fit(
    train_data,
    hyperparameters={
        "MyCustomModel": {},
    },
)
```
If you additionally did 2iii, you can just train with the chosen presets, for example
```python
predictor.fit(
    train_data,
    presets="medium_quality",
)
```

**Reply by Paandaman** [reply]
Hi! 

Thanks a lot for the thorough reply! That does seem quite straight-forward, I'll give it a shot :) 

Thanks again for the reply and for developing AutoGluon!

**Reply by jmren168** [reply]
Hi @Paandaman,

Do you add a new custom model sucessfully? Could you show the steps or share the code via github? Many thanks.

**Reply by ZohrahV** [reply]
Hi @Paandaman, were you able to implement the custom model? I am trying to do the same. But don't want to spend too much time on it.

**Comment by shchur** [comment]
An update as of AutoGluon v1.0.0:
1. You need to implement the model as a class inheriting from `AbstractTimeSeriesModel` and implement the `_fit` and `_predict` methods.
2. You can pass the custom model class directly through `hyperparameters` as
```python
class MyCustomModel(AbstractTimeSeriesModel):
    # ...

predictor.fit(
    train_data,
    hyperparameters={
        MyCustomModel: {},
    },
)

```

**Reply by crossxwill** [reply]
Could you add an example to the [Forecasting Time Series - In Depth](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-indepth.html) tutorial?

I would be interested to see an example of adding this custom model to the autogluon:

https://www.statsmodels.org/stable/generated/statsmodels.tsa.ardl.UECM.html

**Comment by shchur** [comment]
We have added a tutorial describing the process for adding a custom forecasting model: https://auto.gluon.ai/dev/tutorials/timeseries/advanced/forecasting-custom-model.html. Note that it's currently available in the `dev/` version of the documentation.
